File: src/Defender.Portal/src/WebUI/ClientApp/src/models/myGarage/CarRequests.ts
import type { Currency, DateOnly, HistoryType } from "./CarModels";

export interface CreateVehicleRequest {
  displayName: string;
  make: string;
  model: string;
  year: number;
  plate: string;
  vin?: string | null;
}

export type UpdateVehicleRequest = CreateVehicleRequest;

export interface CreateMaintenanceItemRequest {
  name: string;
  intervalMonths?: number | null;
  intervalThousandKm?: number | null;
  lastDate?: DateOnly | null;
  lastOdometerKm?: number | null;
  manualBaselineDate?: DateOnly | null;
  manualBaselineOdometerKm?: number | null;
}

export type UpdateMaintenanceItemRequest = CreateMaintenanceItemRequest;

export interface CreateServiceHistoryRequest {
  date: DateOnly;
  odometerKm: number;
  type: HistoryType;
  title: string;
  notes?: string | null;
  linkedMaintenanceItemIds?: string[];
  costAmountMinor: number | null;
  costCurrency: Currency | null;
}

export type UpdateServiceHistoryRequest = CreateServiceHistoryRequest;

export interface CreateInsurancePolicyRequest {
  provider: string;
  policyNumber?: string | null;
  coverageType?: string | null;
  startDate: DateOnly;
  endDate: DateOnly;
  notes?: string | null;
}

export type UpdateInsurancePolicyRequest = CreateInsurancePolicyRequest;

File: src/Defender.Portal/src/Application/Models/ApiRequests/MyGarage/CarRequests.cs
using Defender.Portal.Application.DTOs.MyGarage;
using Defender.Portal.Application.Enums;

namespace Defender.Portal.Application.Models.ApiRequests.MyGarage;

public sealed class CreateVehicleRequest
{
    public string DisplayName { get; init; } = string.Empty;

    public string Make { get; init; } = string.Empty;

    public string Model { get; init; } = string.Empty;

    public int Year { get; init; }

    public string Plate { get; init; } = string.Empty;

    public string? Vin { get; init; }
}

public sealed class UpdateVehicleRequest
{
    public string DisplayName { get; init; } = string.Empty;

    public string Make { get; init; } = string.Empty;

    public string Model { get; init; } = string.Empty;

    public int Year { get; init; }

    public string Plate { get; init; } = string.Empty;

    public string? Vin { get; init; }
}

public sealed class CreateMaintenanceItemRequest
{
    public string Name { get; init; } = string.Empty;

    public int? IntervalMonths { get; init; }

    public int? IntervalThousandKm { get; init; }

    public DateOnly? LastDate { get; init; }

    public long? LastOdometerKm { get; init; }
}

public sealed class UpdateMaintenanceItemRequest
{
    public string Name { get; init; } = string.Empty;

    public int? IntervalMonths { get; init; }

    public int? IntervalThousandKm { get; init; }

    public DateOnly? LastDate { get; init; }

    public long? LastOdometerKm { get; init; }
}

public sealed class CreateServiceHistoryRequest
{
    public DateOnly Date { get; init; }

    public long OdometerKm { get; init; }

    public HistoryType Type { get; init; }

    public string Title { get; init; } = string.Empty;

    public string? Notes { get; init; }

    public IReadOnlyList<Guid>? LinkedMaintenanceItemIds { get; init; } = [];

    public long? CostAmountMinor { get; init; }

    public Currency? CostCurrency { get; init; }
}

public sealed class UpdateServiceHistoryRequest
{
    public DateOnly Date { get; init; }

    public long OdometerKm { get; init; }

    public HistoryType Type { get; init; }

    public string Title { get; init; } = string.Empty;

    public string? Notes { get; init; }

    public IReadOnlyList<Guid>? LinkedMaintenanceItemIds { get; init; } = [];

    public long? CostAmountMinor { get; init; }

    public Currency? CostCurrency { get; init; }
}

public sealed class CreateInsurancePolicyRequest
{
    public string Provider { get; init; } = string.Empty;

    public string? PolicyNumber { get; init; }

    public string? CoverageType { get; init; }

    public DateOnly StartDate { get; init; }

    public DateOnly EndDate { get; init; }

    public string? Notes { get; init; }
}

public sealed class UpdateInsurancePolicyRequest
{
    public string Provider { get; init; } = string.Empty;

    public string? PolicyNumber { get; init; }

    public string? CoverageType { get; init; }

    public DateOnly StartDate { get; init; }

    public DateOnly EndDate { get; init; }

    public string? Notes { get; init; }
}
